lib/devicetrust/enroll/enroll.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package enroll
+
+import (
+	"context"
+	"runtime"
+
+	"github.com/gravitational/trace"
+
+	devicepb "github.com/gravitational/teleport/api/gen/proto/go/teleport/devicetrust/v1"
+	"github.com/gravitational/teleport/lib/devicetrust/native"
+)
+
+// vars below are used to fake OSes and switch implementations for tests.
+var (
+	getOSType     = getDeviceOSType
+	enrollInit    = native.EnrollDeviceInit
+	signChallenge = native.SignChallenge
+)
+
+// RunCeremony performs the client-side device enrollment ceremony.
+func RunCeremony(ctx context.Context, devicesClient devicepb.DeviceTrustServiceClient, enrollToken string) (*devicepb.Device, error) {
+	// Start by checking the OSType, this lets us exit early with a nicer message
+	// for non-supported OSes.
+	if getOSType() != devicepb.OSType_OS_TYPE_MACOS {
+		return nil, trace.BadParameter("device enrollment not supported for current OS")
```

Comment:
Should we include the detected OS in the error message, just to help with troubleshooting if people run in to this error?

Reply:
Done.